### PR TITLE
Control TCP memory usage in GCS library.

### LINKED ISSUE #2813: Download via storage::Client::ReadObject() does not finish, error, or progress.
A customer reports that sometimes the downloads fail to make progress, with the following traces:


```
$ strace -ttt -p 28864 2>&1 | head -n 30
Process 28864 attached
1560630449.797041 restart_syscall(<... resuming interrupted call ...>) = 0
1560630449.797242 nanosleep({0, 1000000}, 0x7ffc2bc087c0) = 0
1560630449.798477 poll([{fd=22, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
1560630449.798597 poll([{fd=22, events=POLLIN}], 1, 1) = 0 (Timeout)
1560630449.799807 nanosleep({0, 1000000}, 0x7ffc2bc087c0) = 0
1560630449.801022 poll([{fd=22, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
1560630449.801112 poll([{fd=22, events=POLLIN}], 1, 1) = 0 (Timeout)
1560630449.802329 nanosleep({0, 1000000}, 0x7ffc2bc087c0) = 0
1560630449.803598 poll([{fd=22, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
1560630449.803801 poll([{fd=22, events=POLLIN}], 1, 1) = 0 (Timeout)
1560630449.805009 nanosleep({0, 1000000}, 0x7ffc2bc087c0) = 0
1560630449.806264 poll([{fd=22, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
1560630449.806470 poll([{fd=22, events=POLLIN}], 1, 1) = 0 (Timeout)
1560630449.807739 nanosleep({0, 1000000}, 0x7ffc2bc087c0) = 0
1560630449.809041 poll([{fd=22, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
1560630449.809141 poll([{fd=22, events=POLLIN}], 1, 1) = 0 (Timeout)
1560630449.810410 nanosleep({0, 1000000}, 0x7ffc2bc087c0) = 0
1560630449.811640 poll([{fd=22, events=POLLIN|POLLPRI|POLLRDNORM|POLLRDBAND}], 1, 0) = 0 (Timeout)
1560630449.811854 poll([{fd=22, events=POLLIN}], 1, 1) = 0 (Timeout)
1560630449.813107 nanosleep({0, 1000000}, 0x7ffc2bc087c0) = 0
```

```
$ ls -al /proc//28864/fd/22
[snip] /proc//28864/fd/22 -> socket:[213043]
```

```
$ lsof | grep 213043
snip...   22u     IPv4 213043       0t0       TCP [censored customer machine]->74.125.124.128:https (ESTABLISHED)
```

```
#0  0x00002b11459cc00d in nanosleep () from /lib64/libpthread.so.0                                                                                                                                                                                                                
#1  0x00002b116f54440b in std::__1::this_thread::sleep_for(std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000000000l> > const&) () from [censored]/lib/libc++.so.1
#2  0x00002b116f379442 in void std::__1::this_thread::sleep_for<long long, std::__1::ratio<1l, 1000l> >(std::__1::chrono::duration<long long, std::__1::ratio<1l, 1000l> > const&) ()
  from [censored]/lib/libstorage_client.so.1                                                                                                                                                                          
#3  0x00002b116f376036 in google::cloud::storage::v1::internal::CurlDownloadRequest::WaitForHandles(int&) () from [censored]/lib/libstorage_client.so.1
#4  0x00002b116f3747a0 in google::cloud::v0::Status google::cloud::storage::v1::internal::CurlDownloadRequest::Wait<google::cloud::storage::v1::internal::CurlDownloadRequest::Read(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&)::$_1>(g
oogle::cloud::storage::v1::internal::CurlDownloadRequest::Read(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&)::$_1&&) () from [censored]/lib/libstorage_client.so.1
#5  0x00002b116f373a3d in google::cloud::storage::v1::internal::CurlDownloadRequest::Read(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) ()
  from [censored]/lib/libstorage_client.so.1
#6  0x00002b116f439c1d in google::cloud::storage::v1::internal::RetryObjectReadSource::Read(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >&) ()
  from [censored]/lib/libstorage_client.so.1
#7  0x00002b116f4122dd in google::cloud::storage::v1::internal::ObjectReadStreambuf::Peek() () from [censored]/lib/libstorage_client.so.1
#8  0x00002b116f4127c4 in google::cloud::storage::v1::internal::ObjectReadStreambuf::underflow() () from [censored]/lib/libstorage_client.so.1
#9  0x00002b116f4f301a in std::__1::basic_streambuf<char, std::__1::char_traits<char> >::uflow() () from [censored]/lib/libc++.so.1
#10 0x00002b116f4f2fca in std::__1::basic_streambuf<char, std::__1::char_traits<char> >::xsgetn(char*, long) () from [censored]/lib/libc++.so.1
#11 0x00002b116f4f612d in std::__1::basic_istream<char, std::__1::char_traits<char> >::read(char*, long) () from [censored]/lib/libc++.so.1
```

### PR DESCRIPTION
**Pull request by coryan**
The application may need to control the amount of buffers used by
each TCP/IP socket: some applications may need to perform multiple
concurrent downloads, keeping several `storage::ReadObjectStream`
objects open and round-robing over these objects downloading a portion
of the data at a time.

With the default settings on Linux, the socket underlying these objects can
consume several MiB of kernel memory. When the kernel has too many
TCP buffers in use it seems that the kernel stop ACKing data in one or more
sockets to reduce the memory pressure. If one of those sockets is the one
where the application is blocked on `stream.read()` or a similar operation the
application can stop making progress.

Such applications may want to tune the socket parameters to reduce the
memory pressure as they keep multiple sockets open.

This fixes #2813

---
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/2902)

### CODE REVIEW

File: google/cloud/storage/tests/read_object_stall_regression/read_object_stall_test.cc [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+                << r.headers().size();
+      }
+      summary << "\n";
+      ++index;
+    }
+    auto metadata =
+        client.InsertObject(flag_dst_bucket_name, summary_name, summary.str());
+    EXPECT_STATUS_OK(metadata);
+  } while (has_open);
+  std::cout << "DONE: All files closed" << std::endl;
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/tests/read_object_stall_regression/read_object_stall_test.cc]: Fixed.

File: google/cloud/storage/tests/read_object_stall_regression/read_object_stall_test.cc [hunk 2]
```diff
@@ -0,0 +1,306 @@
...
+
+  void AddSomeObjects(Client client) {
+    auto const min_object_size = 32 * 1024 * 1024L;
+    auto const max_object_size = 64 * 1024 * 1024L;
+    std::string contents = MakeRandomData(max_object_size);
+    std::uniform_int_distribution<long> object_size(min_object_size,
+                                                    max_object_size);
+
+    int const object_count = 16;
+    std::vector<std::string> object_names(16);
```
[the remark this answers is not in the document]
Comment (coryan) [reply google/cloud/storage/tests/read_object_stall_regression/read_object_stall_test.cc]: Fixed.

**Comment by coryan** [comment]
> Just a couple of nits, but my big question is:
> 
> > some applications may need to perform multiple parallel downloads
> 
> > If one of those sockets is the one where the application is blocked
> 
> "Parallel" and "the one where the application is blocked" seem to be at odds.

Yes.  Actually the application opens multiple downloads, but round-robins through them, and (given the API) it blocks on each step of the round-robing until N bytes are received.  I changed the description, let me know if it makes better sense now.

> If the application really was draining all readers/sockets in parallel, then the circular wait would not occur, right?

Correct.

> That is, why is the answer "limit per-socket buffer space" rather than "actively drain all your readers"?

For context, the customer is performing a merge sort (basically) across multiple objects in GCS. Sometimes only one of the downloads has data of interesting (consider a merge sort where one of the data sources has a lot more data "early"), so the application is really interested in one particular data source for (potentially) long periods of time.

Alternatively, the application could open a download to just read the portion of the object they want to process in each round-robin step. That workaround works, but it is less efficient: the library needs to start a new download for each round-robin step, and that makes the overall download about 8 times longer.

Or we could change the APIs (but this I would rather not do) to allow select-like behavior when downloading multiple streams. I am not sure that would work anyway, as the problem is the kernel (it seems intentionally) not ACKing messages.

**Comment by coryan** [comment]
PTAL.

**Comment by devbww** [comment]
OK.  Thanks.

I'm happy we'll have some client options that will provide a little bit of control in such situations, and which hopefully can practically address the customer's situation, but those options don't sound like a panacea.

If the user demands that one drainer receives N bytes before any other drainer is serviced, then they're setting themselves up for the possibility of a circular wait.  In as much as the library API contributes towards producing such scenarios, we should think about alternatives.

**Pull request merged**